plugins/techdocs/src/home/components/TechDocsCustomHome.tsx

```diff
@@ -125,6 +125,7 @@ export const TechDocsCustomHome = ({
 
   const { value: entities, loading, error } = useAsync(async () => {
     const response = await catalogApi.getEntities({
+      filter: { 'metadata.annotations.backstage.io/techdocs-ref': '' },
```

Comment:
@benjdlambert @reviewers We as TechDocs folk would naturally love this kinda thing.  But I can also see cases where you might want to be able to differentiate between "this key strictly is an empty string" and "this key is set" and will leave it up to you on how you want to handle that in your API.

Reply:
Yeah, I've not really been over this, but i'd say it could be worth exploring the idea of having `filter` also support an array or set of keys that should be defined rather than this, as you're right it's hard to distinguish between the two

Reply:
So supporting these things involves several different places in the code and protocols right.

There's the `CatalogApi`, as in the commented line here. If we intend to keep the `filter` parameter as a simple key-value object, I would probably prefer that `x: ''` means equal to the empty string, and e.g. `x: CatalogApi.EXISTS` which is a Symbol means to check for existence. Something like that?

Then there's the REST API. I think that the syntax `?filter=x=` should continue to mean equal to the empty string, while e.g. `?filter=x` could mean that it's defined and has any value. Maybe?

Then there's the [`EntityFilter`](https://github.com/backstage/backstage/blob/e47f2a3c095227245e5ac0048daa6fbea989ea8b/plugins/catalog-backend/src/database/types.ts#L129) that this all gets translated to. I would expect that [`matchValueIn`](https://github.com/backstage/backstage/blob/e47f2a3c095227245e5ac0048daa6fbea989ea8b/plugins/catalog-backend/src/database/types.ts#L120) would be amended with a `matchValueExists: true` or something. That would also be possible to extend in the future with a `matchValueExists: false` which requires that it's NOT present.

Then there's the database layer, where this all gets [translated](https://github.com/backstage/backstage/blob/e47f2a3c095227245e5ac0048daa6fbea989ea8b/plugins/catalog-backend/src/next/NextEntitiesCatalog.ts#L79) to actual SQL.

Reply:
Amending with an extra `matchValueExists` prop seems like it might lead to some ambiguity. 

If `matchValueIn` is defined, then by definition, it means the associated key must exist but if `matchValueExists = true` and `matchValueIn` specifies an array of values, does that translate to just checking if the key exists or that it must also match one of those values? and then theres the case where `matchValueExists = false` but `matchValueIn` is still defined, which one takes precedence? A sensible logic flow might be to first check if `matchValueExists` is defined and query based off that ignoring `matchValueIn` and if it's undefined, fallback to the current behavior querying based on `matchValueIn`.

Alternatively, it might be simpler to go with: if `matchValueIn` is defined, the key must exist and match any of the values and `matchValueIn = []` just checks for existence regardless of value. If `matchValueIn` is undefined, the key must not exist. 

Thoughts?

Reply:
It would just redundantly check both. I.e. there would be a SQL fragment that matches for existence and one that matches for values. I'm not concerned about that though, it's a corner case and not harmful. I'd prefer that, over "abusing" the mechanism for values here too. The reader of the code will then have to be aware at all times that "oh ok if the array is empty then i need to make some special case handling of that" since the SQL will definitely be looking different.

Reply:
hmm, if I understand correctly, would this be the proper way to implement this in the following code? https://github.com/backstage/backstage/blob/e47f2a3c095227245e5ac0048daa6fbea989ea8b/plugins/catalog-backend/src/next/NextEntitiesCatalog.ts#L79-L104
```diff
-       for (const { key, matchValueIn } of singleFilter.allOf) { 
+       for (const { key, matchValueIn, matchValueExists } of singleFilter.allOf) {
          // NOTE(freben): This used to be a set of OUTER JOIN, which may seem to
          // make a lot of sense. However, it had abysmal performance on sqlite
          // when datasets grew large, so we're using IN instead.
          const matchQuery = db<DbSearchRow>('search')
            .select('entity_id')
            .where(function keyFilter() {
              this.andWhere({ key: key.toLowerCase() });
-            if (matchValueIn) { 
+            if (matchValueExists !== false && matchValueIn) {
                if (matchValueIn.length === 1) {
                  this.andWhere({ value: matchValueIn[0].toLowerCase() });
                } else if (matchValueIn.length > 1) {
                  this.andWhere(
                    'value',
                    'in',
                    matchValueIn.map(v => v.toLowerCase()),
                  );
                }
              }
            });
-         this.andWhere('entity_id', 'in', matchQuery); 
+         this.andWhere('entity_id', matchValueExists === false ? 'not in' : 'in', matchQuery);
        }
```
taking into account that either `matchValueIn` and `matchValueExists` may be undefined and handling the different combinations of how these two props may be defined. 


Also for completeness, the implementation for the alternative solution already has the empty array case handled inherently and would only require changing the following line. One could also argue that handling of the empty array case is already ambiguous in what the intended behavior should be and requires some special interpretation
```ts
          this.andWhere('entity_id', matchValueIn ? 'in' : 'not in', matchQuery);
```

Reply:
Hm, not in a great position to review in detail now but yes that looks like it's sensible.

Don't worry too much about the edge case of an empty array. Remember that this is populated from parsing a query, and there's no way of having that parser result in a filter set with thar behavior.